### DISCUSSION TITLE
matching fonts from page.get_text("dict") to fitz.Font object

### DISCUSSION THREAD

**Discussion by alexander-fried** [pymupdf/PyMuPDF#4104] [Looking for help]
When extracting font labels for specific text indicated in `page.get_text("dict")`, those labels are different from the fitz.Font objects extracted by first getting the `xref` from `Page.get_fonts()`, then getting the font content data `buffer` from `Document.extract_font(xref)`, and lastly defining the `fitz.Font(fontbuffer=buffer)`.  How does one map the fontnames recovered from the former to a specific `fitz.Font` object?

**Comment by JorjMcKie** [comment]
There is no way to do this securely.
* Nothing forces a PDF creator to put a name in the `page.get_fonts()` items that matches the font's internal name. It could be completely different or some spaces might have been replaced by hyphens, e.g. "FiraGo Regular" vs. "FiraGo-Regular".
* The default in text extraction does not include any font subset prefix like `"ABCDEF+"`. One can request to include these prefixes by setting a global value - see the `pymupdf.TOOLS` documentation.
* MuPDF's `font.name` extraction is a string of no more than 32 bytes. The real internal name may be longer.
* Some fonts cannot be extracted, e.g. Base14 and Type 3 fonts. Their names in text extraction (`span["font"]`) may be synthesized.
* A fontfile binary may not contain any name at all - it is not a mandatory information.

**Reply by alexander-fried** [reply]
What about matching fonts based on the glyph sizes?  Then I can at least get equally-sized fonts, which is good enough for me.  I can get the dimensions of each character with get_text("rawdict") (normalized by size) and I can get the dimensions of the glyphs from the fitz.Font object.  Only trouble with that is that sometimes the dimensions yielded by `fitz.Font.glyph_bbox` are way over-sized.  Which of these glyph properties is something I can trust?

The other thing I can do is write the characters to a new pdf using the known fitz.Font and then remeasure the character sizes using `fitz.get_text("rawdict")`.

Any suggestions, thoughts or caveats regarding the above approaches?

**Reply by JorjMcKie** [reply]
Sounds difficult / complex.
Many times one can guess a suitable font by looking at the name. E.g. if some text particles occur "times", "arial", "courier" etc.

The rawdict output in any case should deliver correct character width values. Glyph bboxes in the font are notoriously unreliable.
